pkg/metrics/clustermesh.go

```diff
@@ -0,0 +1,238 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2022 Authors of Cilium
+
+package metrics
+
+import (
+	"regexp"
+	"sync/atomic"
+	"time"
+
+	clientPkg "github.com/cilium/cilium/pkg/client"
+
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	subsystemClusterMesh = "clustermesh"
+)
+
+var _ prometheus.Collector = (*clustermeshCollector)(nil)
+
+var (
+	// sample remote cluster etcd status
+	// etcd: 1/1 connected, lease-ID=7c027e1caebe4533, lock lease-ID=7c027e1caebe4535, has-quorum=true: https://kind-cilium-mesh-1.mesh.cilium.io:31484 - 3.4.13 (Leader)
+	clusterMeshRemoteClusterEtcdStatusRegex = regexp.MustCompile(`^etcd: \d+/\d+ connected, lease-ID=([0-9a-f]+), lock lease-ID=([0-9a-f]+), has-quorum=true`)
+)
+
+type clustermeshCollector struct {
+	daemonHealthGetter daemonHealthGetter
+
+	remoteClusterEtcdHasQuorumDesc               *prometheus.Desc
+	remoteClusterEtcdTotalObservedLeasesDesc     *prometheus.Desc
+	remoteClusterEtcdTotalObservedLockLeasesDesc *prometheus.Desc
+	remoteClusterLastFailureTimestampDesc        *prometheus.Desc
+	remoteClusterReadinessStatusDesc             *prometheus.Desc
+	remoteClusterTotalFailuresDesc               *prometheus.Desc
+	remoteClusterTotalNodesDesc                  *prometheus.Desc
+	totalGlobalServicesDesc                      *prometheus.Desc
+	totalRemoteClustersDesc                      *prometheus.Desc
+
+	etcdObservedLeaseCnt     int64
+	etcdObservedLockLeaseCnt int64
+
+	lastEtcdLeaseIDs     map[string]string
+	lastEtcdLockLeaseIDs map[string]string
+}
+
+func newClustermeshCollector() *clustermeshCollector {
+	ciliumClient, err := clientPkg.NewClient("")
```

Comment:
All of the information that is being gathered here should already be present in the exact same running binary as where this is coming from, but the code is currently querying the agent's own API in order to gather statistics.

I wonder whether we could structure this a little differently, to just directly push the metrics from the code that tracks or generates these metrics in the first place. For instance, `c.totalRemoteClustersDesc` is generated from the Cilium API clustermesh -> clusters logic, which lives under `pkg/clustermesh`.

There are other examples in the tree of this sort of approach, like in `pkg/ipam/metrics/metrics.go`.